plugin/hover.py

```diff
@@ -106,22 +106,41 @@ def run_async() -> None:
             if not listener:
                 return
             if not only_diagnostics:
-                self.request_symbol_hover_async(listener, hover_point)
-            self._diagnostics_by_config, covering = listener.diagnostics_touching_point_async(
-                hover_point, userprefs().show_diagnostics_severity_level)
-            if self._diagnostics_by_config:
-                self.show_hover(listener, hover_point, only_diagnostics)
-            if not only_diagnostics and userprefs().show_code_actions_in_hover:
-                actions_manager.request_for_region_async(
-                    self.view, covering, self._diagnostics_by_config,
-                    functools.partial(self.handle_code_actions, listener, hover_point))
+                self.request_symbol_hover_async(listener, point_or_region)
+            if isinstance(point_or_region, int):
+                self._diagnostics_by_config, covering = listener.diagnostics_touching_point_async(
+                    point_or_region, userprefs().show_diagnostics_severity_level)
+            else:
+                self._diagnostics_by_config, covering = listener.diagnostics_intersecting_region_async(
+                    point_or_region)
+            if isinstance(point_or_region, int):
+                hover_point = point_or_region
+                if self._diagnostics_by_config:
+                    self.show_hover(listener, hover_point, only_diagnostics)
+                if not only_diagnostics and userprefs().show_code_actions_in_hover:
+                    actions_manager.request_for_region_async(
+                        self.view, covering, self._diagnostics_by_config,
+                        functools.partial(self.handle_code_actions, listener, hover_point))
 
         sublime.set_timeout_async(run_async)
 
-    def request_symbol_hover_async(self, listener: AbstractViewListener, point: int) -> None:
+    def request_symbol_hover_async(self, listener: AbstractViewListener, point_or_region: PointOrRegion) -> None:
         hover_promises = []  # type: List[Promise[ResolvedHover]]
-        document_position = text_document_position_params(self.view, point)
+        point = point_or_region if isinstance(point_or_region, int) else point_or_region.begin()
         for session in listener.sessions_async('hoverProvider'):
+            range_hover_provider = session.get_capability('experimental.rangeHoverProvider')
```

Comment:
Sorry, but I still think this is way too complicated for its own good?

I was imagining a server capability path `hoverProvider.rangeSupport`, which is a boolean, which is `true` when the server has support for a range in the request.

Client side request:

- If the server has `hoverProvider.rangeSupport`, and
- If the position is contained in the current selection, and
- If the current selection is not empty,
- then add `params["range"] = region_to_range(first_selection_region(view), view)`

(I'm unsure if all these additional checks other than the `hoverProvider.rangeSupport` are needed. I could imagine other servers wanting to have the information about the current position in any case, whether it's contained in the range or not)

Server side:

- Check whether the request contains a `"range"` and do something clever with that

Client side response:

- Show whatever the server returns

This whole scheme requires no more than 8 lines of diff, no? We're in a position to cook up an additional server capability and enhancement to the `textDocument/hover` request and I think the above scheme is the least intrusive for the spec.

Reply:
By the way, whether the server capability is considered experimental or not (`experimental.rangeHoverProvider`) is an implementation detail for LSP-metals I think.

Reply:
```
Client side request:

If the server has hoverProvider.rangeSupport, and
If the position is contained in the current selection, and
If the current selection is not empty,
then add params["range"] = region_to_range(first_selection_region(view), view)
(I'm unsure if all these additional checks other than the hoverProvider.rangeSupport are needed. I could imagine other servers wanting to have the information about the current position in any case, whether it's contained in the range or not)
```
Two scenarios are missing:
 - `lsp_hover` is invoked via keyboard and the mouse isn't hovering. `textDocument/hover` should be sent with selection
 -  there is a selection but mouse is hovering an other location. `textDocument/hover` should be sent with mouse position
 
 This is why there are four cases

Reply:
I mean, maybe I want to see both the expression type of the range as well as type info of the position outside of the range using the mouse? Why must that be excluded?

Reply:
yes and it is possible with the current implementation. We just can't send a request with both fields set: https://scalameta.org/metals/docs/integrations/new-editor#textdocumenthover

>  Why must that be excluded?
it is not excluded but as I said there are multiple combinations and these four checks are there for that. 
Are you saying you can handle all those cases with fewer checks ?

Reply:
> `lsp_hover` is invoked via keyboard and the mouse isn't hovering. `textDocument/hover` should be sent with selection

This still feels wrong to me and I would argue to not send the selection if it is entirely unrelated to mouse hover. Shouldn't it be a separate command which can be used to assign a key binding then? For example `lsp_selection_info` or so.

Reply:
Well, actually the `lsp_hover` already takes the optional `point` argument, and as implemented here it is just extending this concept. So feel free to ignore my concern if it's useful to combine them all into `lsp_hover`, even though not really related to "hover".

Reply:
> > `lsp_hover` is invoked via keyboard and the mouse isn't hovering. `textDocument/hover` should be sent with selection
> 
> This still feels wrong to me and I would argue to not send the selection if it is entirely unrelated to mouse hover. Shouldn't it be a separate command which can be used to assign a key binding then? For example `lsp_selection_info` or so.

When using the keyboard the "point" has to follow the caret. It wouldn't make sense to follow the mouse pointer when using keyboard.

Yes, the naming of `textDcoument/hover` request makes it confusing to think about in terms of keyboard usage but if we want to allow triggering the hover info from the keyboard (which we are and should) then it only makes sense to follow the caret position. It would be unusable otherwise.

Reply:
Right, I think I had this confusion because I don't have this keybinding set up, and because `lsp_hover` does different things depending on how it was invoked. Should be fine then as it is!

Reply:
@rwols I simplified it now and all cases work. Sorry for missing your point last time.